File: mobile/appium-python/utils/smart_find.py
import time
from appium.webdriver.common.appiumby import AppiumBy
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import (
    TimeoutException, NoSuchElementException, StaleElementReferenceException
)


class SmartFind:
    """Self-healing element finder for mobile — tries accessibility-id → resource-id → xpath."""

    def __init__(self, driver, timeout=10):
        self.driver = driver
        self.timeout = timeout

    def find(self, accessibility_id=None, resource_id=None, xpath=None):
        per = self.timeout / 3

        if accessibility_id:
            try:
                return WebDriverWait(self.driver, per).until(
                    EC.visibility_of_element_located((AppiumBy.ACCESSIBILITY_ID, accessibility_id))
                )
            except (TimeoutException, NoSuchElementException):
                pass

        if resource_id:
            try:
                return WebDriverWait(self.driver, per).until(
                    EC.visibility_of_element_located((AppiumBy.ID, resource_id))
                )
            except (TimeoutException, NoSuchElementException):
                pass

        if xpath:
            return WebDriverWait(self.driver, per).until(
                EC.visibility_of_element_located((AppiumBy.XPATH, xpath))
            )

        raise NoSuchElementException("SmartFind: all strategies failed")

    def retry(self, action, attempts=3, delay=0.5):
        """Retry action with stale element protection."""
        for i in range(attempts):
            try:
                return action()
            except StaleElementReferenceException:
                if i == attempts - 1:
                    raise
                time.sleep(delay * (i + 1))

File: mobile/appium-java/src/test/java/utils/SmartFind.java
package utils;

import io.appium.java_client.AppiumDriver;
import org.openqa.selenium.*;
import org.openqa.selenium.support.ui.*;
import java.time.Duration;

/**
 * Self-healing element finder for mobile — tries accessibility-id → resource-id → xpath with retry.
 */
public class SmartFind {
    private final AppiumDriver driver;
    private final Duration timeout;

    public SmartFind(AppiumDriver driver) {
        this(driver, Duration.ofSeconds(10));
    }

    public SmartFind(AppiumDriver driver, Duration timeout) {
        this.driver = driver;
        this.timeout = timeout;
    }

    public WebElement find(String accessibilityId, String resourceId, String xpath) {
        Duration per = timeout.dividedBy(3);

        if (accessibilityId != null) {
            try {
                return new WebDriverWait(driver, per)
                    .until(ExpectedConditions.visibilityOfElementLocated(
                        AppiumBy.accessibilityId(accessibilityId)));
            } catch (TimeoutException ignored) {}
        }
        if (resourceId != null) {
            try {
                return new WebDriverWait(driver, per)
                    .until(ExpectedConditions.visibilityOfElementLocated(
                        AppiumBy.id(resourceId)));
            } catch (TimeoutException ignored) {}
        }
        if (xpath != null) {
            return new WebDriverWait(driver, per)
                .until(ExpectedConditions.visibilityOfElementLocated(By.xpath(xpath)));
        }
        throw new NoSuchElementException("SmartFind: all strategies failed");
    }

    /** Retry action with stale element protection */
    public <T> T retry(java.util.function.Supplier<T> action, int attempts) {
        for (int i = 0; i < attempts; i++) {
            try {
                return action.get();
            } catch (StaleElementReferenceException e) {
                if (i == attempts - 1) throw e;
                try { Thread.sleep(500L * (i + 1)); } catch (InterruptedException ignored) {}
            }
        }
        throw new RuntimeException("retry exhausted");
    }
}
